### PR TITLE
Support Non-printable ASCII characters in AWS S3 metadata

### PR DESCRIPTION
**Pull request by thiagodasilva**
When syncing metadata to s3, encode non-ascii characters
and when migrating metadata from s3, make sure to decode.

[Fixes #156776841]

### CODE REVIEW

File: s3_sync/utils.py [hunk 1]
```diff
@@ -845,6 +848,24 @@ def convert_to_s3_headers(swift_headers):
...
+    if value.startswith('=?utf-8?q?'):
+        return '=?UTF-8?Q?' + value[10:]
+    elif value.startswith('=?utf-8?b?'):
+        return '=?UTF-8?B?' + value[10:]
+    else:
+        return value
+
+
+def decode_s3_header(value):
+    header, charset = decode_header(value)[0]
```
Comment (timuralp) [inline s3_sync/utils.py]: Should we check the returned `charset` value?

File: s3_sync/utils.py [hunk 2]
```diff
@@ -845,6 +848,27 @@ def convert_to_s3_headers(swift_headers):
...
+    if charset not in ('utf-8', 'UTF-8', 'utf8'):
```
Comment (mdegerne) [inline s3_sync/utils.py]: Out of curiosity, how did you come up with this specific list of permutations?

Comment (thiagodasilva) [reply s3_sync/utils.py]: TBH just out of experience in our own code, not a very empirical method. utf8 should actually not be used according to standards[1], but I see it used all the time :man_shrugging: 

[1] - https://en.wikipedia.org/wiki/UTF-8#Official_name_and_variants

File: s3_sync/utils.py (hunk 1 above)
Comment (thiagodasilva) [reply s3_sync/utils.py]: done, added the check

File: s3_sync/utils.py [hunk 3]
```diff
@@ -854,15 +878,15 @@ def convert_to_swift_headers(s3_headers):
             swift_headers[header[len(S3_USER_META_PREFIX):]] = value
         elif header.startswith(S3_USER_META_PREFIX):
             key = SWIFT_USER_META_PREFIX + header[len(S3_USER_META_PREFIX):]
-            swift_headers[key] = value
+            swift_headers[key] = decode_s3_header(value)
```
Comment (timuralp) [inline s3_sync/utils.py]: Please add unit tests to cover the mime-encoding conversion.

File: s3_sync/utils.py [hunk 4]
```diff
@@ -845,6 +848,27 @@ def convert_to_s3_headers(swift_headers):
     return s3_headers
 
 
+def sanitize_s3_header(value):
```
Comment (timuralp) [inline s3_sync/utils.py]: Can you add unit tests for this function and/or expand the unit tests for `convert_to_s3_headers()`?

File: s3_sync/utils.py [hunk 5]
```diff
@@ -845,6 +848,27 @@ def convert_to_s3_headers(swift_headers):
...
+        return '=?UTF-8?B?' + value[10:]
+    else:
+        return value
+
+
+def decode_s3_header(value):
+    header, charset = decode_header(value)[0]
+    if charset not in ('utf-8', 'UTF-8', 'utf8'):
+        return value
+    else:
```
Comment (timuralp) [inline s3_sync/utils.py]: Can drop the else and just return header.

**Pull request merged**